### DISCUSSION TITLE
My full instructions for doing this on Windows with Home Assistant's ESPHome add-on

### DISCUSSION THREAD

**Discussion by den-mac** [emporia-vue-local/esphome#195] [General]
So there's a lot of info here on how to do this, all based around using a Windows python tool called esphome.py. I had no luck trying to get it all going, but was able to use the tool to backup the factory firmware, which I recommend you do.

To do that I followed the instructions on how to install python and the esphome tool. Then I ran this to do the backup:

`esptool -b 921600 read_flash 0 0x800000 flash_contents.bin`

Note for me the tool is just called esptool, not esptool.py. When installing Python BE SURE to click the "add to path" checkbox on the install dialog so the tool will work.

==============

Flashing procedure:

1. Assemble USB setup and get it ready to go for backing up the factory f/w and flashing to ESPhome.

Some notes for my particular setup:

- The particular adapter I used ([Amazon link](https://www.amazon.com/dp/B07BBPX8B8)) labels the RX/TX backwards, so on mine RX<->RX and TX<->TX. Also for Windows you have to install a driver. I got mine here: https://ftdichip.com/drivers/

- You have to use the 5V pin and supply 5V to the unit. It does not work with the 3.3v only.

- You'll want to ground the IO0 for flashing. Note that I used a breadboard too split my ground from the USB adapter into two clips that attached to GND and IO0 on the Vue2.

2. In ESPHome, create a new device and give it a name. Then EDIT and copy the api key and OTA key somewhere.

3. Use the template provided on (on [this page](https://github.com/emporia-vue-local/esphome#writing-configuration)) and update the two items from above.

Other notes:

- for the sensor section, you can rename the CT labels with your actual names. Also update the "Daily Power" section at the bottom with the same names.

- Be sure to double check the "phase_a" or "phase_b" to ensure your particular branch circuit is on that side of the panel. For me the odd ones were on phase_a and the even ones were on phase_b.

- For 240V circuits, or double sized breakers that use both A+B, be sure to add the "multiply: 2" part on the circuit so you get the proper wattage for your device. I only had one of these, my A/C condenser.

- Finally, I have two Vue2 units and only one has it's A+B main CT coils connected. I think I'll need to comment out some things in the 2nd Vue that doesn't have anything connected to it's A/B jacks but I haven't gotten there yet.

- Note on WIFI - be sure your "Secrets" section of ESPHome (upper right) is updated with your wifi into. Also, you CANNOT have a hidden SSID for ESP devices to work; figure that out the hard way. One other note, I am unable to set up a DHCP reservation for ESPhome devices; no idea what's going on there. My reservations work for everything else, just not ESP home devices. Dunno. I'll look into how to set up a static IP in the config. I think this is how to do static IPs:

Add just under the wifi SSID + key part:

```
  manual_ip:
    static_ip: 192.168.x.x
    gateway: 192.168.x.x
    subnet: 255.255.255.0
```

4. Build a firmware image via "Install" and "Manual install -> Modern format"

5. If you run into build problems, try to "Clean Build Files" and try again. I have a Rpi 4 and ran into memory problems a couple times but upon attempting the build again it would work for me. Most recently I needed to do a simple "Restart -> YAML restart" to get it to build.

6. Once you get a successful bin file, we need to flash it to the Vue. I use Firefox which doesn't support web USB, so I have to launch Chrome and go to [https://web.esphome.io/?dashboard_wizard](https://web.esphome.io/?dashboard_wizard)

7. Click Install then choose the image file, then flash it! Hopefully this part works for you and it will take a little time.

8. Once it's done, unplug the USB adapter and remove the wire grounding IO0 so it will come up normally. Replug the USB to power up the board with the new firmware.

9. After the device boots, go to your ESPHome dashboard in HA and ensure it shows your device "online."

10. Go to Integrations and it should show the new device there ready to be added, which will add the entities for your circuits.

11. Once you see all the entities and such, you're done and can reassemble the Vue2 and put it back into your electrical panel!

Good luck and I hope this helps *someone* out there as it was extremely difficult for me but this procedure worked for me.

Dennis

P.S. I want to give a sincere THANK YOU to the community and folks that worked on this project. It's awesome that I can use my Emporia devices locally now, as I was running into troubles with the api not updating my power data as often as I would like.

**Comment by den-mac** [comment]
![IMG_3509](https://github.com/emporia-vue-local/esphome/assets/95404985/6500b881-74ae-449f-baff-de0ee15c1d6b)
The whole setup

![IMG_3511](https://github.com/emporia-vue-local/esphome/assets/95404985/c6abec2b-6d9f-462d-b1ac-d0e8acb7a859)
My particular jumper setup. Note the clips for connecting both GND and the IO0 pin for flashing.

![IMG_3512](https://github.com/emporia-vue-local/esphome/assets/95404985/449e7e00-99f9-4938-b071-27d46a32fb6b)
My USB adapter has a voltage jumper. Be sure to select 5V.

![IMG_3508](https://github.com/emporia-vue-local/esphome/assets/95404985/7180a5e7-8d4f-4b76-a0df-2ef16a2265f6)
I recommend using a solder sucker to remove the existing solder, then soldering in a header as it makes things easier.

**Comment by den-mac** [comment]
I have confirmed the way to set a static ip is under the wifi section. Add this in:

```
manual_ip:
    static_ip: 192.168.x.x
    gateway: 192.168.x.x
    subnet: 255.255.255.0
```

Also, if you happen to fall into the same situation as me where the DHCP server already assigned a dynamic IP to your device, adding the above and trying to install wirelessly will fail since it will try to contact using the new static IP.

Just add this line after the above to get it to use the old IP to flash:

```
  use_address: 192.168.22.218
```

So my full wifi config to wirelessly push to the device is something like this:

```
wifi:
  ssid: *****
  password: *****

  manual_ip:
    static_ip: 192.168.2.142
    gateway: 192.168.2.1
    subnet: 255.255.255.0
  
  use_address: 192.168.2.218
```

**Reply by Shesakillatwo** [reply]
First let me say your instructions were awesome and worked perfectly.  I am very new to ESP32 programming and to ESPHome.  I did get through the entire setup where you stated I can install this back into my panel.  I however have not installed this in the panel yet as I wanted to get it all set up before doing that.  So, I obviously do not know some of the specific details such as what clamps will be on which circuits and their phases once installed.  

I am planning to install this in the panel this weekend when I will gather some of the details I do not yet have.   so I wanted to check and confirm that I can simply update the "myvue2.yaml" file in Home assistant and re-run the install once I have inserted all the details and it should update to the correct values?  I don't really want to disassemble this again to uppdate and reload the detaailed data.  Please let me know!  Thanks again for all your work on this.....

**Comment by den-mac** [comment]
Yes, once you put it all back together the future updates will occur over wifi via ESPHome in HA.

This happens often as the ESPhome software gets updated every now and then and you have to update all your ESPHome devices as well. They're all done over wifi.

Keep one thing in mind (I don't remember if I mentioned it above) - if you use DHCP at first to get your device online, and then set up a static reservation or static IP later, you'll need to do something special in the yaml file to accommodate the old IP (which is what it will use to connect) and the new static config, which will be applied after it restarts.

But again, that's all done over wifi and you'll never need to pull the board out unless you need to reflash it again (I have not needed to do this myself on both of my units).

**Reply by Darkyputz** [reply]
Hello den...
I just found your instructions after failing so many times already with flashing this that i almost gave up and called it a paper weight.
I did with 3.3volt, hope thats the issue that i uave.
And just to be clear. Do i have to have ground and io0 connected the whole time or io0 just short at the beginning. In that fact so many instructions are vague.
Thx for any tip. And i use the same flasher. Have u tried to use ctr and rst on the flasher ?

**Reply by den-mac** [reply]
Just ground for startup (when you apply power) then I release the button. Once it's in bootloader mode it stays there until a reboot.

I have not tried using any of the "Clear to send (CTS) or clear to receive (CTR)" pins, nor the reset pin.

**Reply by Darkyputz** [reply]
Thx for answering.
Looks like something is broken on mine, I am at a point now where it works. Butonly when on flasher with 5v, rx, tx, cts and ctr connected. In that combination, the emporia boots, connects to wifi and everything, even ota update works. But as soon i unplug ctr/cts or plug it into original power supply, o ly the power light goes. Do they have a bootlaoder that can be dead and the flasher is mimicking that missing functionality?
Hope u have an idea

**Comment by supermarkert** [comment]
> The particular adapter I used ... labels the RX/TX backwards, so on mine RX<->RX and TX<->TX. 

FYI, I believe it's the Vue boards that are (intentionally) labeled backwards. My guess is it's for convenience, so you would know to match your RX<-->RX and TX<->TX terminals from your adapter. But on the Vue board itself, the Rx terminal is connected to the Tx pin on the ESP. So every adapter should be matching Rx to Rx terminals, etc.
